Sources/PackageRegistry/RegistryManager.swift

```diff
@@ -0,0 +1,296 @@
+                   let data = response.body
+                {
+                    let fileSystem = InMemoryFileSystem()
+
+                    let filename: String
+                    if let swiftLanguageVersion = swiftLanguageVersion {
+                        filename = Manifest.basename + "@swift-\(swiftLanguageVersion).swift"
+                    } else {
+                        filename = Manifest.basename + ".swift"
+                    }
+
+                    try fileSystem.writeFileContents(.root.appending(component: filename), bytes: ByteString(data))
+                    manifestLoader.load(
+                        at: .root,
+                        packageIdentity: package.identity,
+                        packageKind: .registry(package.identity),
+                        packageLocation: package.location,
+                        version: version,
+                        revision: nil,
+                        toolsVersion: .currentToolsVersion,
+                        identityResolver: self.identityResolver,
+                        fileSystem: fileSystem,
+                        diagnostics: self.diagnosticEngine,
+                        on: .sharedConcurrent,
+                        completion: completion
+                    )
+                } else {
+                    throw RegistryError.invalidResponse
+                }
+            } catch {
+                queue.async {
+                    completion(.failure(error))
+                }
+            }
+        }
+    }
+
+    public func downloadSourceArchive(
+        for version: Version,
+        of package: PackageReference,
+        into fileSystem: FileSystem,
+        at destinationPath: AbsolutePath,
+        expectedChecksum: ByteString? = nil,
+        on queue: DispatchQueue,
+        completion: @escaping (Result<Void, Error>) -> Void
+    ) {
+        guard case let (scope, name)? = package.identity.scopeAndName else {
+            return completion(.failure(RegistryError.invalidPackage(package)))
+        }
+
+        guard let registry = configuration.registry(for: scope) else {
+            return completion(.failure(RegistryError.registryNotConfigured(scope: scope)))
+        }
+
+        var components = URLComponents(url: registry.url, resolvingAgainstBaseURL: true)
+        components?.appendPathComponents("\(scope)", "\(name)", "\(version).zip")
+
+        guard let url = components?.url else {
+            return completion(.failure(RegistryError.invalidURL))
+        }
+
+        var request = HTTPClient.Request(
+            method: .get,
+            url: url,
+            headers: [
+                "Accept": "application/vnd.swift.registry.v1+zip"
+            ]
+        )
+
+        request.options.authorizationProvider = authorizationProvider
+
+        client.execute(request, progress: nil) { result in
+            switch result {
+            case .success(let response):
+                if response.statusCode == 200,
```

Comment:
iirc one can configure the request with acceptable response codes to avoid such check, tho you do make other checks here as well

Reply:
Yeah, if it were just the status check or a range of status codes, I would've gone with that. But it's less code overall to do it this way here.